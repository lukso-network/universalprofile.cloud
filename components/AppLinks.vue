<script setup lang="ts">
import { detectSocialMedia } from '@lukso/web-components/tools'

import type { LinkMetadata } from '@lukso/lsp-smart-contracts'

type Props = {
  links?: LinkMetadata[]
}

type Slots = {
  default(props: {
    hasLinks: boolean
    socialMediaLinks: LinkMetadataSocialMedia[]
    otherLinks: LinkMetadataSocialMedia[]
    linksParsed: LinkMetadataSocialMedia[]
  }): any
}

const props = defineProps<Props>()
defineSlots<Slots>()

const linksParsed = computed(
  () =>
    props?.links
      ?.filter(link => link.url)
      ?.map(
        link =>
          ({
            ...link,
            socialMediaName: detectSocialMedia(link?.url),
          }) as LinkMetadataSocialMedia
      ) || []
)

const hasLinks = computed(() => !!linksParsed.value?.length)

const socialMediaLinks = computed(() =>
  linksParsed.value?.filter(link => !!link.socialMediaName)
)
const otherLinks = computed(() =>
  linksParsed.value?.filter(link => !link.socialMediaName)
)
</script>

<template>
  <slot
    :has-links="hasLinks"
    :social-media-links="socialMediaLinks"
    :other-links="otherLinks"
    :links-parsed="linksParsed"
  />
</template>
