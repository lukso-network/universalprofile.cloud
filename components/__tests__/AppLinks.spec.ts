import { renderSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'

import AppLinks from '@/components/AppLinks.vue'

describe('AppLinks', () => {
  it('should render empty links', async () => {
    const component = await renderSuspended(AppLinks, {
      props: {
        links: [],
      },
      slots: {
        default:
          '<template #default="{ socialMediaLinks }">{{socialMediaLinks}}</template>',
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render socialMediaLinks', async () => {
    const component = await renderSuspended(AppLinks, {
      props: {
        links: [
          {
            title: 'Twitter',
            url: 'https://twitter.com',
          },
          {
            title: 'Custom link',
            url: 'https://example.com',
          },
        ],
      },
      slots: {
        default:
          '<template #default="{ socialMediaLinks }">{{socialMediaLinks}}</template>',
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render otherLinks', async () => {
    const component = await renderSuspended(AppLinks, {
      props: {
        links: [
          {
            title: 'Twitter',
            url: 'https://twitter.com',
          },
          {
            title: 'Custom link',
            url: 'https://example.com',
          },
        ],
      },
      slots: {
        default:
          '<template #default="{ otherLinks }">{{otherLinks}}</template>',
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render hasLinks', async () => {
    const component = await renderSuspended(AppLinks, {
      props: {
        links: [
          {
            title: 'Twitter',
            url: 'https://twitter.com',
          },
          {
            title: 'Custom link',
            url: 'https://example.com',
          },
        ],
      },
      slots: {
        default: '<template #default="{ hasLinks }">{{hasLinks}}</template>',
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render linksParsed', async () => {
    const component = await renderSuspended(AppLinks, {
      props: {
        links: [
          {
            title: 'Twitter',
            url: 'https://twitter.com',
          },
          {
            title: 'Custom link',
            url: 'https://example.com',
          },
        ],
      },
      slots: {
        default:
          '<template #default="{ linksParsed }">{{linksParsed}}</template>',
      },
    })

    expect(component.html()).toMatchSnapshot()
  })
})
