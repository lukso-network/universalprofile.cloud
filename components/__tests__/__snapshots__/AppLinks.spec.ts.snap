// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppLinks > should render empty links 1`] = `"<div id="test-wrapper">[]</div>"`;

exports[`AppLinks > should render hasLinks 1`] = `"<div id="test-wrapper">true</div>"`;

exports[`AppLinks > should render linksParsed 1`] = `
"<div id="test-wrapper">[
  {
  "title": "Twitter",
  "url": "https://twitter.com",
  "socialMediaName": "x"
  },
  {
  "title": "Custom link",
  "url": "https://example.com"
  }
  ]</div>"
`;

exports[`AppLinks > should render otherLinks 1`] = `
"<div id="test-wrapper">[
  {
  "title": "Custom link",
  "url": "https://example.com"
  }
  ]</div>"
`;

exports[`AppLinks > should render socialMediaLinks 1`] = `
"<div id="test-wrapper">[
  {
  "title": "Twitter",
  "url": "https://twitter.com",
  "socialMediaName": "x"
  }
  ]</div>"
`;
